Kafka source #90

sobychacko opened the pull request:
resolves #90

Commit:
Initial commit for porting Kafka source module into Spring Cloud Stream modules

Basic Kafka source

Minimum set of properties needed for the source

Initial set of dependency changes in mvn

Commit:
Changes to Kafka source module

Commit:
Kafka source integration tests

sobychacko committed:
simplifying the Kafka source integrtaiton test producer

sobychacko committed:
Polishing Kafka source:

Integration tests
Multiple partition tests
Adding initial offset support as a spring boot property

markfisher commented on kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaSourceApplication.java:
```diff
@@ -0,0 +1,34 @@
+/*
```
some funky formatting for this header comment

sobychacko replied:
Just updated the PR with formatting changes.

sobychacko committed:
Fixing copyright formatting in class files

ilayaperumalg commented on kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaConfigurationProperties.java:
```diff
@@ -0,0 +1,122 @@
+ * @author Soby Chacko
+ */
+@ConfigurationProperties
+public class KafkaConfigurationProperties {
```
There are few other configuration properties such as `concurrency`, `maxFetch`, `queueSize` and `stopTimeout` of `KafkaMessageListenerContainer` can be defined here.

sobychacko replied:
@ilayaperumalg I am adding these 4 properties. Any other properties that you can think of? I think we can start with these and add more later on as the need arises.

ilayaperumalg replied:
@sobychacko yeah, and those are the only properties I found missing as well.

ilayaperumalg commented on kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaSource.java:
```diff
@@ -0,0 +1,153 @@
+	}
+
+	private Decoder<?> getPayloadDecoder() throws ClassNotFoundException, IllegalAccessException, InstantiationException {
+		String payloadDecoder = properties.getKeyDecoder();
```
properties.getPayloadDecoder() here.

ilayaperumalg commented on kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaSource.java:
```diff
@@ -0,0 +1,153 @@
+				.get();
+	}
+
+	private Decoder<?> getKeyDecoder() throws ClassNotFoundException, IllegalAccessException, InstantiationException {
```
Can we refactor both of these `getKeyDecoder()` and `getPayloadDecoder()` to `getDecoder(String decoderType)` as both look functionally the same?

sobychacko replied:
@ilayaperumalg the decoders may not be the same. It was a typo that we used the KeyDecoder for both key and payload. Thanks a lot for catching this. We have separate properties for key and payload decoders in the KafkaConfigurationProperties class. I will change the code to reflect that. 
For example, Key could be a LongDecoder and the payload could be based on String. With that said, we can refactor some common elements between these two methods.

sobychacko replied:
@ilayaperumalg with that said, I think we can re-factor some common elements from these two methods.

ilayaperumalg commented on pom.xml:
```diff
@@ -110,6 +111,11 @@
 				</exclusions>
 			</dependency>
 			<dependency>
+				<groupId>org.springframework.integration</groupId>
```
I am not sure why do we need this in parent POM. May be this can be removed with just the version tag on the kafka source POM.

sobychacko replied:
@ilayaperumalg sure thing - i pulled it from the parent and put it in the module itself.

ilayaperumalg commented on kafka-source/README.adoc:
````diff
@@ -0,0 +1,24 @@
+== Run
+
+```
+$> java -jar target/kafka-source-${version}-exec.jar
````
It would be good to include the doc on how would someone specify the `partitions` and `initialOffsets` options when running the module here.

For instance with options like these:

```
--topics=test123,test234 --partitions.test123=1,2 --partitions.test234=1
```

sobychacko replied:
Good point, will address this.

ilayaperumalg commented on kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaSource.java:
```diff
@@ -0,0 +1,153 @@
+		return IntegrationFlows.from(Kafka.messageDriverChannelAdapter(container())
+				.keyDecoder(getKeyDecoder())
+				.payloadDecoder(getPayloadDecoder()))
+				.channel(source.output())
```
Can this be simplified to use the channel name `Source.OUTPUT` directly instead of `source.output()` since the `IntegrationFlows` builder supports specifying the channel name in this method?

ericbottard replied:
IMO, using the channel reference is better. Even if using `Source.OUTPUT` would be refactoring-friendly, it's just some attribute of the channel. What we want to convey is that we want _that_ channel, wouldn't you agree?

ilayaperumalg replied:
I just tried running the source module exec with the `topics` option set and I see the following when using `source.output()` here. But this works ok when having the channel name `Source.OUTPUT`. Looks like it has got something to do with the channel name being assigned for the module output channel `output` vs 'application.fromKafka.channel#0'.?

```
org.springframework.messaging.MessageDeliveryException: Dispatcher has no subscribers for channel 'application.fromKafka.channel#0'.; nested exception is org.springframework.integration.MessageDispatchingException: Dispatcher has no subscribers
    at org.springframework.integration.channel.AbstractSubscribableChannel.doSend(AbstractSubscribableChannel.java:81) ~[spring-integration-core-4.2.4.RELEASE.jar!/:na]
    at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:442) ~[spring-integration-core-4.2.4.RELEASE.jar!/:na]
    at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:392) ~[spring-integration-core-4.2.4.RELEASE.jar!/:na]
    at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:115) ~[spring-messaging-4.2.4.RELEASE.jar!/:4.2.4.RELEASE]
    at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:45) ~[spring-messaging-4.2.4.RELEASE.jar!/:4.2.4.RELEASE]
    at org.springframework.messaging.core.AbstractMessageSendingTemplate.send(AbstractMessageSendingTemplate.java:105) ~[spring-messaging-4.2.4.RELEASE.jar!/:4.2.4.RELEASE]
    at org.springframework.integration.endpoint.MessageProducerSupport.sendMessage(MessageProducerSupport.java:105) ~[spring-integration-core-4.2.4.RELEASE.jar!/:na]
    at org.springframework.integration.kafka.inbound.KafkaMessageDrivenChannelAdapter.access$300(KafkaMessageDrivenChannelAdapter.java:43) ~[spring-integration-kafka-1.3.0.RELEASE.jar!/:na]
    at org.springframework.integration.kafka.inbound.KafkaMessageDrivenChannelAdapter$AutoAcknowledgingChannelForwardingMessageListener.doOnMessage(KafkaMessageDrivenChannelAdapter.java:171) ~[spring-integration-kafka-1.3.0.RELEASE.jar!/:na]
    at org.springframework.integration.kafka.listener.AbstractDecodingMessageListener.onMessage(AbstractDecodingMessageListener.java:50) ~[spring-integration-kafka-1.3.0.RELEASE.jar!/:na]
    at org.springframework.integration.kafka.listener.QueueingMessageListenerInvoker$KafkaMessageDispatchingSubscriber.onNext(QueueingMessageListenerInvoker.java:221) [spring-integration-kafka-1.3.0.RELEASE.jar!/:na]
    at org.springframework.integration.kafka.listener.QueueingMessageListenerInvoker$KafkaMessageDispatchingSubscriber.onNext(QueueingMessageListenerInvoker.java:209) [spring-integration-kafka-1.3.0.RELEASE.jar!/:na]
    at reactor.core.processor.util.RingBufferSubscriberUtils.route(RingBufferSubscriberUtils.java:67) [reactor-core-2.0.7.RELEASE.jar!/:na]
    at reactor.core.processor.RingBufferProcessor$BatchSignalProcessor.run(RingBufferProcessor.java:789) [reactor-core-2.0.7.RELEASE.jar!/:na]
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_20]
    at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_20]
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_20]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_20]
    at java.lang.Thread.run(Thread.java:745) [na:1.8.0_20]
Caused by: org.springframework.integration.MessageDispatchingException: Dispatcher has no subscribers
    at org.springframework.integration.dispatcher.UnicastingDispatcher.doDispatch(UnicastingDispatcher.java:153) ~[spring-integration-core-4.2.4.RELEASE.jar!/:na]
    at org.springframework.integration.dispatcher.UnicastingDispatcher.dispatch(UnicastingDispatcher.java:120) ~[spring-integration-core-4.2.4.RELEASE.jar!/:na]
    at org.springframework.integration.channel.AbstractSubscribableChannel.doSend(AbstractSubscribableChannel.java:77) ~[spring-integration-core-4.2.4.RELEASE.jar!/:na]
    ... 18 common frames omitted
```

mbogoevici replied:
+1 on using channel reference

sobychacko committed:
Kafka source polishing

More configuration properties
decoder method refactorin
more examples in doc
Add kafka dependency only in the kafka-source module

sobychacko commented:
Addressed the feedback of the review. I haven't changed the output channel from reference to Source.OUTPUT as we probably need to discuss this a bit further it seems like.

sobychacko committed:
Kafka source refactoring

Kafka source cleanup/refactoring

Add more integration tests to cover partitions, offsets etc.

sobychacko commented:
@mbogoevici Few more changes need to be committed for the source.

artembilan commented on kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaSource.java:
```diff
@@ -0,0 +1,158 @@
+	public KafkaMessageListenerContainer container() {
+
+		Assert.isTrue(!ObjectUtils.isEmpty(properties.getTopics()) || !ObjectUtils.isEmpty(properties.getPartitions()),
+				"Either a list of topics OR a topic=<comma separated partitions> must be provided");
```
The `Assert` must like `validation` in the `KafkaConfigurationProperties`.
From other side this message doesn't reflect `topics/partitions` reality.

sobychacko replied:
@artembilan Can you please elaborate on this as to what you meant?

artembilan replied:
I mean something like this:

```
@AssertTrue(message = "exactly one of  'topics' and 'partitions' are mutually exclusive")
public boolean isTopicsOrPartitions() {
    return !ObjectUtils.isEmpty(getTopics()) || !ObjectUtils.isEmpty(getPartitions();
}
```

and place it into the `KafkaConfigurationProperties` and here in the `@Configuration` just do `if...else`.

Although I don't know: maybe we should use something as a default topic?..

sobychacko replied:
+1 on the validation stuff. 

I don't think we should default to any topic. We should let the users provide the topic that they want to consume from. If the topic is not present or the topics are not given, then we should simply error out. For the sink case, defining a default topic might make sense.

artembilan commented on kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaSource.java:
```diff
@@ -0,0 +1,158 @@
+	}
+
+	@Bean
+	public OffsetManager kafkaNativeOffsetManager(ConnectionFactory kafkaConnectionFactory) {
```
I don't see the `ConnectionFactory` `@Bean`.
Who populates it?
What have I missed? Why then `container()` bean doesn't use the same injection technique?

artembilan commented on kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaSource.java:
```diff
@@ -0,0 +1,158 @@
+				new KafkaMessageListenerContainer(kafkaConnectionFactory, properties.getTopics()) :
+				new KafkaMessageListenerContainer(kafkaConnectionFactory, partitions);
+
+		container.setOffsetManager(kafkaNativeOffsetManager(kafkaConnectionFactory));
```
Oh! I see now. Interesting... :smile: 
Re. `ConnectionFactory` `@Bean`.
Actually the `DefaultConnectionFactory` should really be as a `@Bean`

sobychacko replied:
@artembilan good point. Will change that as a bean

artembilan commented on kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaConfigurationProperties.java:
```diff
@@ -0,0 +1,159 @@
+
+	private String keyDecoder;
+
+	private String payloadDecoder;
```
I see that these are converted to the `Class` in the `KafkaSource` thereafter.
The `private Class payloadDecoder;` works well here in the properties.

artembilan commented on kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaConfigurationProperties.java:
```diff
@@ -0,0 +1,159 @@
+	private int concurrency;
+	private int maxFetch;
+	private int stopTimeout;
+	private int queueSize;
```
Blank lines in between class members.
Minor of course, but safe the time during review :smile:

artembilan commented on kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaConfigurationProperties.java:
```diff
@@ -0,0 +1,159 @@
+		this.partitions = partitions;
+	}
+
+	public String getZkConnect() {
```
`@NotBlank`

sobychacko replied:
@artembilan these are optional values and for zk connect, there is a default value that we fall back to.

artembilan replied:
Well, I mean something like this:

```
@IntegrationTest({"topics = foo", "zkConnect="})
```

when we just can't fall back, because of end-user interaction.

artembilan commented on kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaConfigurationProperties.java:
```diff
@@ -0,0 +1,159 @@
+		this.zkConnect = zkConnect;
+	}
+
+	public String getZkConnectionTimeout() {
```
`@NotBlank`

artembilan commented on kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaConfigurationProperties.java:
```diff
@@ -0,0 +1,159 @@
+		this.zkConnectionTimeout = zkConnectionTimeout;
+	}
+
+	public String getZkSessionTimeout() {
```
`@NotBlank`

artembilan commented on kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaSource.java:
```diff
@@ -0,0 +1,158 @@
+	}
+
+	@Bean
+	public ZookeeperConnect zookeeperConnect() {
```
I'd say that `ZookeeperConfiguration` should be as a bean, not this one.

artembilan commented on kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaSource.java:
```diff
@@ -0,0 +1,158 @@
+	@Bean
+	public IntegrationFlow fromKafka() throws IllegalAccessException, InstantiationException, ClassNotFoundException {
+
+		return IntegrationFlows.from(Kafka.messageDriverChannelAdapter(container())
```
Having such a simple flow, I'd say that there is no reason with the Java DSL overhead.

artembilan commented on kafka-source/src/test/java/org/springframework/cloud/stream/module/kafka/source/KafkaSourceIntegrationTests.java:
```diff
@@ -0,0 +1,306 @@
+@DirtiesContext
+public abstract class KafkaSourceIntegrationTests {
+
+	private static String scs_kafka_test_embedded;
```
I don't like underscore names in Java. Just IMO...
No reason to react, if you think different and leave it as is.

artembilan commented on kafka-source/src/test/java/org/springframework/cloud/stream/module/kafka/source/KafkaSourceIntegrationTests.java:
```diff
@@ -0,0 +1,306 @@
+import static scala.collection.JavaConversions.asScalaBuffer;
+
+/**
+ * Tests for basic {@link KafkaSource}.
```
`basic`? Is there (or planned) an `intermediate` or even `advance`? :smile: 

Just a nit-pick, of course. I can live with this anyway.

artembilan commented on kafka-source/src/test/java/org/springframework/cloud/stream/module/kafka/source/KafkaSourceIntegrationTests.java:
```diff
@@ -0,0 +1,306 @@
+import java.util.*;
+import java.util.concurrent.TimeUnit;
+
+import static org.junit.Assert.*;
```
No one asterisk import.
And some wrong order:
![imports](https://cloud.githubusercontent.com/assets/1071590/12902704/b7c0e948-ce90-11e5-8915-1e5e92a5c7a9.jpg)

artembilan commented on kafka-source/src/test/java/org/springframework/cloud/stream/module/kafka/source/KafkaSourceIntegrationTests.java:
```diff
@@ -0,0 +1,306 @@
+	public static void cleanup() {
+		if (scs_kafka_test_embedded != null) {
+			System.setProperty("SCS_KAFKA_TEST_EMBEDDED", scs_kafka_test_embedded);
+		} else {
```
`else` from the new line.
As well as `catch`, `finally` etc..

artembilan commented on kafka-source/src/test/java/org/springframework/cloud/stream/module/kafka/source/KafkaSourceIntegrationTests.java:
```diff
@@ -0,0 +1,306 @@
+	}
+
+	@Configuration
+	@EnableIntegration
```
This stuff is redundant for the configuration below.
Just chasing for some test performance. :smile:

artembilan commented on pom.xml:
```diff
@@ -186,7 +187,7 @@
 			<artifactId>spring-cloud-spring-service-connector</artifactId>
 		</dependency>
-		<!--TODO: Move transport related dependencies here pending https://jira.spring.io/browse/XD-3337  -->
+		<!--TODO: Move transporct related dependencies here pending https://jira.spring.io/browse/XD-3337  -->
```
M-m-m. The previous one was better. :smile:

sobychacko committed:
Kafka source polishing

    - Fixing formatting issues
    - Adding more validation in @ConfigurationProperties
    - Adding more tests for properties
    - @Bean fixes in KafkaSource

sobychacko commented:
@artembilan @mbogoevici I tried to address all the comments above. Can you review?

artembilan commented on kafka-source/pom.xml:
```diff
@@ -0,0 +1,69 @@
+        <dependency>
+            <groupId>org.springframework.cloud.stream.module</groupId>
+            <artifactId>spring-cloud-stream-modules-test-support</artifactId>
+            <version>1.0.0.BUILD-SNAPSHOT</version>
```
The `version` is redundant here.
We are in the same project with its modules.

I wonder if we even can avoid explicit `1.0.0.BUILD-SNAPSHOT` in the `<parent>` definition.
For example use some external `properties`, like we have with Gradle...

sobychacko committed:
Kafka source mvn pom cleanup

artembilan commented on kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaConfigurationProperties.java:
```diff
@@ -0,0 +1,181 @@
+import org.springframework.util.ObjectUtils;
+
+/**
+ * Properties related to connecting to Kafka broker.
```
I think this JavaDocs isn't relevant. We connect to ZK with the Kafka cluster behind.
Maybe just something like:

```
* The Kafka Source properties.
```

?

artembilan commented on kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaSource.java:
```diff
@@ -0,0 +1,225 @@
+                new KafkaMessageListenerContainer(kafkaConnectionFactory(), partitions);
+
+        // if we have less target partitions than target concurrency, adjust accordingly
+        container.setConcurrency(Math.min(properties.getConcurrency(), partitions.length));
```
This is redundant exercise. See `ConcurrentMessageListenerDispatcher` ctor:

```
this.consumers = Math.min(partitions.size(), consumers);
```

artembilan replied:
OK. Never mind.
I see now the reason. `ConcurrentMessageListenerDispatcher` has a code like:

```
Assert.isTrue(consumers <= partitions.size(),
                "The number of consumers must be smaller or equal to the number of partitions");
........
this.consumers = Math.min(partitions.size(), consumers);
```

So, we don't reach that `Math.min()` if we provide more consumers than partitions.

I guess that `Assert` should be removed in the SIK. The  `Math.min()` is a good compromise do not disturb end-user.

mbogoevici replied:
We can display a warning, I suppose, and label it `maxConcurrency` or something else. The main problem is that they could set a concurrency level that will never be honoured by the running application.

artembilan replied:
Anyway I think it is a SIK responsibility not here.
Although we should duplicated it here until release there :smile:

artembilan commented on kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaSource.java:
```diff
@@ -0,0 +1,225 @@
+        container.setMaxFetch(properties.getMaxFetch());
+        container.setStopTimeout(properties.getStopTimeout());
+        container.setQueueSize(properties.getQueueSize());
+        container.setOffsetManager(kafkaNativeOffsetManager(kafkaConnectionFactory()));
```
Shouldn't we consider to use the `WindowingOffsetManager` wrapper here as well?
See discussion in the https://github.com/spring-cloud/spring-cloud-stream/pull/316

sobychacko replied:
good point. Will wrap it in Windowing..Manager.

artembilan commented on kafka-source/src/test/java/org/springframework/cloud/stream/module/kafka/source/KafkaConfigurationPropertiesTest.java:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    @Test
+    public void testValidTopicSceanrio() throws Exception {
```
Typo: Sce**na**rio :smile:

artembilan commented on kafka-source/src/test/java/org/springframework/cloud/stream/module/kafka/source/KafkaConfigurationPropertiesTest.java:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    @Test
+    public void testValidPartitionsSceanrio() throws Exception {
```
DITO

artembilan commented on kafka-source/src/test/java/org/springframework/cloud/stream/module/kafka/source/KafkaConfigurationPropertiesTest.java:
```diff
@@ -0,0 +1,118 @@
+        catch (Exception e) {
+            assertThat(e, instanceOf(BeanCreationException.class));
+            assertThat(e.getMessage(),
+                    containsString("Either a list of topics (--topics) OR partitions.<topic>=<comma separated partitions> (--partitions.<topic>) must be provided"));
```
IMO it's better do not overflow 120 code line length.
At least it is easier to review here on GH.

artembilan commented on kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaConfigurationProperties.java:
```diff
@@ -0,0 +1,181 @@
+        this.zkSyncTime = zkSyncTime;
+    }
+
+    public Class getKeyDecoder() {
```
Should be like `Class<?>` to avoid the compiler noise.

artembilan commented on kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaSource.java:
```diff
@@ -0,0 +1,225 @@
+    }
+
+    @Bean
+    public KafkaMessageDrivenChannelAdapter kafkaMessageDrivenChannelAdapter() throws IllegalAccessException, InstantiationException, ClassNotFoundException {
```
The `ClassNotFoundException` is redundant already.

artembilan commented on kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaSource.java:
```diff
@@ -0,0 +1,225 @@
+
+    @Bean
+    public KafkaMessageDrivenChannelAdapter kafkaMessageDrivenChannelAdapter() throws IllegalAccessException, InstantiationException, ClassNotFoundException {
+        final KafkaMessageDrivenChannelAdapter kafkaMessageDrivenChannelAdapter =
```
Since you don't use SI Java DSL anymore, the `spring-integration-java-dsl` dependency can be removed.

sobychacko committed:
Kafka source polishing

Generic compilation warning fix
Remove unused dependency from mvn pom
More logging for max concurrency settings

mbogoevici commented:
As part of the ongoing discussion on what types of OOTB modules we want to provide for messaging systems, the Kafka source and sink are too similar to the Kafka binder in terms of functionality, so we will rely on the binder for supporting this functionality (also on multibinder support if needing to connect to different types of systems, or cluster instances). 

Current gaps in the binder: 
- support for multiple topics/destinations in the binder: https://github.com/spring-cloud/spring-cloud-stream/issues/334
- listening to specific partitions: https://github.com/spring-cloud/spring-cloud-stream/issues/336
- resetting to arbitrary topics: https://github.com/spring-cloud/spring-cloud-stream/issues/335

mbogoevici commented:
Closed as per above comment.
